internal/dataplane/kongstate/kongstate.go

```diff
@@ -690,3 +683,305 @@ func maybeLogKongIngressDeprecationError(logger logr.Logger, services []*corev1.
+	// Note: this procedure will make referred services'/routes'/consumers' ID to be filled.
+	// So it requires the `FillIDs` feature gate to be enabled.
+	for fieldName, field := range schema.Fields {
+		if field.Type != EntityFieldTypeForeign {
+			continue
+		}
+		switch field.Reference {
+		case string(kong.EntityTypeServices):
+			serviceIDs := getServiceIDFromPluginRels(logger, rels, pluginRelEntities.RouteAttachedService, workspace)
+			// TODO: we should generate multile entities if the plugin is attached to multiple services/routes/consumers.
+			// https://github.com/Kong/kubernetes-ingress-controller/issues/6123
+			if len(serviceIDs) > 0 {
+				parsedEntity[fieldName] = map[string]interface{}{
+					"id": serviceIDs[0],
+				}
+				logger.V(util.DebugLevel).Info("added ref to service", "service_id", serviceIDs[0])
+			}
+		case string(kong.EntityTypeRoutes):
+			routeIDs := lo.FilterMap(rels.Routes, func(r *Route, _ int) (string, bool) {
+				if err := r.FillID(workspace); err != nil {
+					return "", false
+				}
+				return *r.ID, true
+			})
+			if len(routeIDs) > 0 {
+				parsedEntity[fieldName] = map[string]interface{}{
+					"id": routeIDs[0],
+				}
+				logger.V(util.DebugLevel).Info("added ref to route", "route_id", routeIDs[0])
+			}
+		case string(kong.EntityTypeConsumers):
+			consumerIDs := lo.FilterMap(rels.Consumers, func(c *Consumer, _ int) (string, bool) {
+				if err := c.FillID(workspace); err != nil {
+					return "", false
+				}
+				return *c.ID, true
+			})
+			if len(consumerIDs) > 0 {
+				parsedEntity[fieldName] = map[string]interface{}{
+					"id": consumerIDs[0],
+				}
+				logger.V(util.DebugLevel).Info("added ref to consumer", "consumer_id", consumerIDs[0])
+			}
+		}
+	}
+}
+
+// getServiceIDFromPluginRels returns the ID of the services which a plugin refers to in RelatedEntitiesRef.
+// It fills the IDs of services directly referred, and IDs of services where referred routes attaches to.
+func getServiceIDFromPluginRels(log logr.Logger, rels RelatedEntitiesRef, routeAttachedService map[string]*Service, workspace string) []string {
+	// Return IDs of directly referred services.
+	if len(rels.Services) > 0 {
+		return lo.FilterMap(rels.Services, func(s *Service, _ int) (string, bool) {
+			if err := s.FillID(workspace); err != nil {
+				log.Error(err, "failed to fill ID for service")
+				return "", false
+			}
+			return *s.ID, true
+		},
+		)
+	}
+	// Returns IDs of services where the referred routes attaches.
+	if len(rels.Routes) > 0 {
+		serviceIDs := lo.FilterMap(
+			rels.Routes, func(r *Route, _ int) (string, bool) {
+				svc, ok := routeAttachedService[*r.Name]
+				if !ok {
+					return "", false
+				}
+				if err := svc.FillID(workspace); err != nil {
+					log.Error(err, "failed to fill ID for service")
+					return "", false
+				}
+				return *svc.ID, true
+			},
+		)
+		return lo.Uniq(serviceIDs)
```

Comment:
You could consider accounting for `serviceIDs` as you go instead of mapping and filtering **and then** taking the unique ones. That would optimize the memory usage and save us from calling `lo.Uniq()`

Reply:
Do you mean calling `addRelation` to add the relationship between plugins and services in traversing routes?